src/tasty/tasty.tsx

```diff
@@ -0,0 +1,215 @@
+import styledComponents, { createGlobalStyle } from 'styled-components';
+import { ComponentType, FC, forwardRef, useContext, useMemo } from 'react';
+import { isValidElementType } from 'react-is';
+import { BreakpointsContext } from './providers/BreakpointsProvider';
+import { modAttrs } from './utils/modAttrs';
+import { AllBaseProps, BaseStyleProps, Props } from './types';
+import { renderStyles } from './utils/renderStyles';
+import { pointsToZones } from './utils/responsive';
+import { Styles, StylesInterface } from './styles/types';
+import { BASE_STYLES } from './styles/list';
+import { ResponsiveStyleValue } from './utils/styles';
+import { mergeStyles } from './utils/mergeStyles';
+
+export type TastyProps<
+  K extends (keyof StylesInterface)[],
+  DefaultProps = Props,
+> = {
+  /** The tag name of the element. */
+  as?: string;
+  /** Default styles of the element. */
+  styles?: Styles;
+  /** The list of styles that can be provided by props */
+  styleProps?: K;
+} & Omit<DefaultProps, 'as' | 'styles' | 'styleProps'>;
+
+export interface GlobalTastyProps {
+  breakpoints?: number[];
+}
+
+export type AllBasePropsWithMods<K extends (keyof StylesInterface)[]> =
+  AllBaseProps & {
+    [key in K[number]]?: ResponsiveStyleValue<StylesInterface[key]>;
+  } & BaseStyleProps;
+
+type TastyPropsWithDefaults<
+  Props extends { styles?: Styles },
+  DefaultProps extends Partial<Props>,
+> = keyof DefaultProps extends never
+  ? Props
+  : {
+      [key in Extract<keyof Props, keyof DefaultProps>]?: Props[key];
+    } & {
+      [key in keyof Omit<Props, keyof DefaultProps>]: Props[key];
+    };
+
+function tasty<K extends (keyof StylesInterface)[]>(
+  options: TastyProps<K>,
+  secondArg?: never,
+);
```

Comment:
I guess we should thoroughly type this component, since this is a core thing of our design system